### PR TITLE
[ready] Clean up torch.distributions

### PR DESCRIPTION
**Pull request by vishwakftw**
This PR removes instances of `Variable` and `Tensor` across the entire distributions suite. I have also removed some dead code, and made minor modifications to `utils.py` inside `distributions`. Tensor notation has also been fixed in the examples.

cc: @fritzo @apaszke

### CODE REVIEW

**Comment by SsnL** [comment]
Didn't look at the details. Just want to say thanks for cleaning the code!!

File: torch/distributions/exp_family.py [hunk 1]
```diff
@@ -52,7 +51,7 @@ def entropy(self):
         Method to compute the entropy using Bregman divergence of the log normalizer.
         """
         result = -self._mean_carrier_measure
-        nparams = [Variable(p.data, requires_grad=True) for p in self._natural_params]
+        nparams = [p.clone().requires_grad_() for p in self._natural_params]
```
Comment (fmassa) [inline torch/distributions/exp_family.py]: Is this equivalent? Don't we need to use something like `detach()` here?

Comment (vishwakftw) [reply torch/distributions/exp_family.py]: I am not quite sure of the semantics of `detach()`, so help would be greatly appreciated!

Comment (SsnL) [reply torch/distributions/exp_family.py]: @vishwakftw Ha! Then the migration guide is your best friend! See the `What about .data?` section of https://pytorch.org/2018/04/22/0_4_0-migration-guide.html .

Comment (vishwakftw) [reply torch/distributions/exp_family.py]: I actually made changes prior to the push based on the migration guide, but ran into this error, which is why I resorted to the existing version. The error is below:
```
RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn
```
However, in the case with `Variable`, the older version: the `_natural_params` where not necessarily `Variable`s with `requires_grad=True`. Hence the older version.

Comment (SsnL) [reply torch/distributions/exp_family.py]: I think @fmassa is talking about using `.detach().requires_grad_()` combined. Did you try that too?

Comment (vishwakftw) [reply torch/distributions/exp_family.py]: Oh I see. I got it working with that. Thanks @fmassa and @SsnL

**Comment by vishwakftw** [comment]
Windows failure seems to be spurious.

**Comment by SsnL** [comment]
Hmm so that error isn’t just on BN... 

@pytorchbot retest this please

**Comment by vishwakftw** [comment]
Could not change this: https://github.com/pytorch/pytorch/pull/8046/commits/1ce7817f238adbfc5246daeb11c03c204f9cd057#diff-ac77b2f5b8a834eb5f3e310c611998e0R64

File: torch/distributions/distribution.py [hunk 2]
```diff
@@ -196,13 +196,13 @@ def _validate_sample(self, value):
...
         Args:
-            value (Tensor): the tensor whose log probability is to be
+            value (tensor): the tensor whose log probability is to be
                 computed by the `log_prob` method.
         Raises
             ValueError: when the rightmost dimensions of `value` do not match the
                 distribution's batch and event shapes.
         """
-        if not isinstance(value, torch.Tensor):
+        if value.__class__.__name__ != 'Tensor':
```
Comment (apaszke) [inline torch/distributions/distribution.py]: The check was correct before. `isinstance(value, torch.Tensor)` is the recommended way now

File: torch/distributions/utils.py [hunk 3]
```diff
@@ -19,29 +19,19 @@
...
     r"""
-    Return floating point info about a `Tensor`:
+    Return floating point info about a `tensor`:
     - `.eps` is the smallest number that can be added to 1 without being lost.
     - `.tiny` is the smallest positive number greater than zero
       (much smaller than `.eps`).
 
     Args:
-        tensor (Tensor): tensor of floating point data.
+        tensor (torch tensor): tensor of floating point data.
```
Comment (apaszke) [inline torch/distributions/utils.py]: Can you please remove `torch`

File: torch/distributions/utils.py [hunk 4]
```diff
@@ -119,11 +102,10 @@ def log_sum_exp(tensor, keepdim=True):
     summing is done along the last dimension.
 
     Args:
-        tensor (torch.Tensor)
+        tensor (torch tensor)
```
Comment (apaszke) [inline torch/distributions/utils.py]: Can you just say `Tensor`? That's what we do everywhere else

**Review by apaszke — COMMENTED** [review]
Some minor issues, and should be ready to merge after that.

File: torch/distributions/distribution.py (hunk 2 above)
Comment (vishwakftw) [reply torch/distributions/distribution.py]: I have reverted the change.

However, considering the deprecation of `torch.Tensor` based on the discussion in https://github.com/pytorch/pytorch/issues/7900 , how would you suggest this check be written?

**Review by apaszke — APPROVED** [review]

File: torch/distributions/distribution.py (hunk 2 above)
Comment (apaszke) [reply torch/distributions/distribution.py]: To be honest I didn't know of that thread, and I don't think there's any clear deprecation plan for `torch.Tensor` as a type (maybe there is as a tensor constructor).

**Comment by apaszke** [comment]
Thanks @vishwakftw!

**Pull request merged**